### PR TITLE
Add a setup script to install Julia + IJulia

### PR DESCRIPTION
**Pull request by yuvipanda**
## Describe your changes

Based on https://github.com/jupyter/docker-stacks/pull/1926#issuecomment-1611914753,
move Julia installation to a common script.

## Checklist (especially for first-time contributors)

- [ ] I have performed a self-review of my code
- [ ] If it is a core feature, I have added thorough tests
- [ ] I will try not to use force-push to make the review process easier for reviewers
- [ ] I have updated the documentation for significant changes

### CODE REVIEW

**Comment by yuvipanda** [comment]
@mathbunnyru So I added the script, but am not sure how exactly to get it into both the julia and datascience notebooks. I've proposed one way here, stolen from @cboettig's ideas in rocker (https://github.com/rocker-org/rocker-versioned2/blob/965e5a8dbb493dfced6f1463ae7f1f5cd53987e5/dockerfiles/r-ver_4.3.1.Dockerfile#L19). This also allows easier extension downstream - for example, if someone wants to add julia to a spark image, for example. They can just extend and call `install-julia.bash`. LMK if this sounds good to you?

**Comment by mathbunnyru** [comment]
> @mathbunnyru So I added the script, but am not sure how exactly to get it into both the julia and datascience notebooks. I've proposed one way here, stolen from @cboettig's ideas in rocker (https://github.com/rocker-org/rocker-versioned2/blob/965e5a8dbb493dfced6f1463ae7f1f5cd53987e5/dockerfiles/r-ver_4.3.1.Dockerfile#L19). This also allows easier extension downstream - for example, if someone wants to add julia to a spark image, for example. They can just extend and call `install-julia.bash`. LMK if this sounds good to you?

I totally agree with this approach and I actually proposed it as well in my comment in a previous PR 😃

**Comment by yuvipanda** [comment]
@mathbunnyru great! waiting for tests to complete, but what do you think of this? :)

File: minimal-notebook/setup-scripts/setup-ijulia.bash [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+# Install base Julia packages
+julia -e <<EOF
+import Pkg;
+Pkg.update();
+Pkg.add("HDF5", "IJulia");
+Pkg.precompile();
+EOF
+
+
```
Comment (mathbunnyru) [inline minimal-notebook/setup-scripts/setup-ijulia.bash:15]: I think we don't need extra empty line here.

**Review by mathbunnyru — APPROVED** [review]
Looks really nice.
I would like to propose to move the Julia version handling to the bash script as well (and remove it from both Dockerfiles)
Something like - if environment variable is set, use it. If not, use the default version defined in the bash script.

**Comment by yuvipanda** [comment]
@mathbunnyru done!

File: datascience-notebook/Dockerfile [hunk 2]
```diff
@@ -30,32 +24,11 @@ RUN apt-get update --yes && \
 # install Julia packages in /opt/julia instead of ${HOME}
 ENV JULIA_DEPOT_PATH=/opt/julia \
     JULIA_PKGDIR=/opt/julia \
-    JULIA_VERSION="${julia_version}"
 
 WORKDIR /tmp
```
Comment (mathbunnyru) [inline datascience-notebook/Dockerfile:28]: It's better to move this to the part where it's needed in the bash script.

File: datascience-notebook/Dockerfile [hunk 3]
```diff
@@ -87,17 +60,7 @@ RUN mamba install --yes \
...
-RUN julia -e 'import Pkg; Pkg.update()' && \
-    julia -e 'import Pkg; Pkg.add("HDF5")' && \
-    julia -e 'using Pkg; pkg"add IJulia"; pkg"precompile"' && \
-    # move kernelspec out of home \
-    mv "${HOME}/.local/share/jupyter/kernels/julia"* "${CONDA_DIR}/share/jupyter/kernels/" && \
-    chmod -R go+rx "${CONDA_DIR}/share/jupyter" && \
-    rm -rf "${HOME}/.local" && \
-    fix-permissions "${JULIA_PKGDIR}" "${CONDA_DIR}/share/jupyter"
+# Setup IJulia kernel
+RUN /usr/local/bin/setup-ijulia.bash
```
Comment (mathbunnyru) [inline datascience-notebook/Dockerfile:64]: Let's move this a little bit up, it makes sense to have it after the installation.

File: julia-notebook/Dockerfile [hunk 4]
```diff
@@ -12,56 +12,19 @@ SHELL ["/bin/bash", "-o", "pipefail", "-c"]
...
-ARG julia_version="1.9.1"
-
 # Julia dependencies
 # install Julia packages in /opt/julia instead of ${HOME}
 ENV JULIA_DEPOT_PATH=/opt/julia \
-    JULIA_PKGDIR=/opt/julia \
-    JULIA_VERSION="${julia_version}"
+    JULIA_PKGDIR=/opt/julia
 
 WORKDIR /tmp
```
Comment (mathbunnyru) [inline julia-notebook/Dockerfile:20]: This one will also be removed.

File: minimal-notebook/Dockerfile [hunk 5]
```diff
@@ -43,3 +43,6 @@ USER ${NB_UID}
 
 # Add R mimetype option to specify how the plot returns from R to the browser
 COPY --chown=${NB_UID}:${NB_GID} Rprofile.site /opt/conda/lib/R/etc/
+
+# Add setup scripts that may be used by downstream images or inherited images
+COPY setup-scripts/* /usr/bin/local/
```
Comment (mathbunnyru) [inline minimal-notebook/Dockerfile:48]: It should be `/usr/local/bin/`, I suppose 🙂
Or, we can use `/opt`.

File: minimal-notebook/setup-scripts/setup-ijulia.bash [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+# - Julia is already set up, with the setup-julia.bash command
+set -exuo pipefail
+
+# Install base Julia packages
+julia -e <<EOF
+import Pkg;
+Pkg.update();
+Pkg.add("HDF5", "IJulia");
+Pkg.precompile();
+EOF
```
Comment (mathbunnyru) [inline minimal-notebook/setup-scripts/setup-ijulia.bash:13]: That's nice 👍

File: minimal-notebook/setup-scripts/setup-julia.bash [hunk 7]
```diff
@@ -0,0 +1,38 @@
+#!/bin/bash
+set -exuo pipefail
+# Requirements:
+# - This script is run as the root user
+# - The JULIA_PKGDIR environment variable is set
+
+# Default julia version to install if env var is not set
+# Check https://julialang.org/downloads/
+JULIA_VERSION="${JULIA_VERSION:-1.9.1}"
```
Comment (mathbunnyru) [inline minimal-notebook/setup-scripts/setup-julia.bash:9]: 👍

**Review by mathbunnyru — COMMENTED** [review]
Looks really good now.
Just a few minor things to cleanup and we can merge this :)

File: minimal-notebook/Dockerfile (hunk 5 above)
Comment (yuvipanda) [reply minimal-notebook/Dockerfile:48]: lolol I spent a while trying to figure out why the test broke :D

File: julia-notebook/Dockerfile (hunk 4 above)
Comment (yuvipanda) [reply julia-notebook/Dockerfile]: Done

File: datascience-notebook/Dockerfile (hunk 3 above)
Comment (yuvipanda) [reply datascience-notebook/Dockerfile]: Done!

File: minimal-notebook/Dockerfile (hunk 5 above)
Comment (yuvipanda) [reply minimal-notebook/Dockerfile]: Moved to /opt

File: datascience-notebook/Dockerfile (hunk 2 above)
Comment (yuvipanda) [reply datascience-notebook/Dockerfile]: Done!

**Comment by yuvipanda** [comment]
hmm, it looks like you need to explicitly approve the GitHub run, @mathbunnyru! Not sure why that happened suddenly.

**Comment by mathbunnyru** [comment]
The last commit was from the `pre-commit.ci[bot]` (run `pre-commit install` in the folder with this repo, your life will be easier 🙂 ) .
And we use self-hosted runners here, they are not isolated, so I only allow actions to be run either by members of the organization or I have to run them explicitly, when there is an outside contributor.

**Comment by yuvipanda** [comment]
Yeah, been relying on pre-commit.ci because I fucked up my local node install bad enough that prettier won't run anymore and am too lazy to fix it :D 

Oooh, TIL this uses self hosted runners! Thank you :)

**Comment by yuvipanda** [comment]
Hooray, tests pass!

**Comment by yuvipanda** [comment]
Yay, thank you @mathbunnyru!

**Pull request merged**